src/backend/crossterm.rs

```diff
@@ -24,36 +23,310 @@ use crate::{
+/// A [`Backend`] implementation that uses [Crossterm] to render to the terminal.
+///
+/// The `CrosstermBackend` struct is a wrapper around a type implementing [`std::io::Write`], which
+/// is used to send commands to the terminal. It provides methods for drawing content, manipulating
+/// the cursor, and clearing the terminal screen.
+///
+/// # Examples
 ///
-/// The `CrosstermBackend` struct is a wrapper around a type implementing `Write`, which
-/// is used to send commands to the terminal. It provides methods for drawing content,
-/// manipulating the cursor, and clearing the terminal screen.
+/// Generally apps should [`CrosstermBackend::on_stderr()`] or [`CrosstermBackend::on_stdout()`]
+/// methods, which both enable raw mode and enter the alternate screen. Choosing `stderr` over
+/// `stdout` ensures your app displays on the terminal even if the standard output stream is
+/// redirected and makes it easy to write apps that can be piped to other programs.
+///
+/// ```rust,no_run
+/// use ratatui::backend::CrosstermBackend;
+///
+/// let mut backend = CrosstermBackend::on_stdout()?;
+/// // alternatively
+/// let mut backend = CrosstermBackend::on_stderr()?;
+/// # std::io::Result::Ok(())
+/// ```
 ///
-/// # Example
+/// For more control over raw mode, the alternate screen, or mouse capture use
+/// [`CrosstermBackend::new()`].
 ///
-/// ```rust
-/// use ratatui::backend::{Backend, CrosstermBackend};
+/// ```rust,no_run
+/// use std::io::stdout;
+/// use ratatui::backend::CrosstermBackend;
 ///
-/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
-/// let buffer = std::io::stdout();
-/// let mut backend = CrosstermBackend::new(buffer);
-/// backend.clear()?;
-/// # Ok(())
-/// # }
+/// let mut backend = CrosstermBackend::new(stdout())
+///     .with_raw_mode()?
+///     .with_alternate_screen()?
+///     .with_mouse_capture()?;
+/// # std::io::Result::Ok(())
 /// ```
-#[derive(Debug, Default, Clone, Eq, PartialEq)]
+///
+/// [Crossterm]: https://docs.rs/crossterm
+#[derive(Debug, Clone, Eq, PartialEq)]
 pub struct CrosstermBackend<W: Write> {
+    /// The buffer being written to.
     buffer: W,
+    /// Whether raw mode is enabled. This is used to determine whether to disable raw mode when
+    /// dropping the backend.
+    raw_mode_enabled: bool,
+    /// Whether the alternate screen is entered. This is used to determine whether to leave the
+    /// alternate screen when dropping the backend.
+    alternate_screen_entered: bool,
+    /// Whether mouse capture is enabled. This is used to determine whether to disable mouse
+    /// capture when dropping the backend.
+    mouse_capture_enabled: bool,
 }
 
 impl<W> CrosstermBackend<W>
 where
     W: Write,
 {
     /// Creates a new `CrosstermBackend` with the given buffer.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use std::io::stdout;
+    /// # use ratatui::backend::CrosstermBackend;
+    /// let mut backend = CrosstermBackend::new(stdout());
```

Comment:
My rationale for including even trivial examples like this is because Ratatui is the sort of crate that first time rust users end up using, so making it as easy as possible at the expense of some verbosity is a reasonable tradeoff.

The secondary thing I like about super simple examples is that they act like a canary for when a method changes signature in an incompatible ways.

Reply:
Hmm. Looking at the rust docs, it does look like they have example docs even for 'trivial' methods, so it's possible I'm just wrong on this. Ultimately this isn't something I feel super passionate about either way. :)

Reply:
Sounds good - I'll wait on other opinions. Leaning towards keeping it as-is.

Reply:
In general, I think more examples are better, especially for people new to rust _and_ new to ratatui. In this particular case, I don't feel strongly one way or the other.